File: python/src/falcon_core/physics/config/core/adjacency.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.adjacency import Adjacency as _CAdjacency
from falcon_core.physics.device_structures.connections import Connections
from falcon_core.generic.f_array import FArray
from falcon_core.generic.list import List
from falcon_core.generic.list import List

class Adjacency:
    """Python wrapper for Adjacency."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, data: Any, shape: Any, ndim: Any, indexes: Connections) -> Adjacency:
        return cls(_CAdjacency.new(data, shape, ndim, indexes._c if indexes is not None else None))

    @classmethod
    def from_json(cls, json: str) -> Adjacency:
        return cls(_CAdjacency.from_json(json))

    def indexes(self, ) -> Connections:
        ret = self._c.indexes()
        if ret is None: return None
        return Connections._from_capi(ret)

    def indexes(self, ) -> Connections:
        ret = self._c.indexes()
        if ret is None: return None
        return Connections._from_capi(ret)

    def get_true_pairs(self, ) -> List:
        ret = self._c.get_true_pairs()
        if ret is None: return None
        return List(ret)

    def size(self, ) -> None:
        ret = self._c.size()
        return ret

    def dimension(self, ) -> None:
        ret = self._c.dimension()
        return ret

    def shape(self, out_buffer: Any, ndim: Any) -> None:
        ret = self._c.shape(out_buffer, ndim)
        return ret

    def data(self, out_buffer: Any, numdata: Any) -> None:
        ret = self._c.data(out_buffer, numdata)
        return ret

    def times_equals_farray(self, other: FArray) -> None:
        ret = self._c.times_equals_farray(other._c if other is not None else None)
        return ret

    def times_farray(self, other: FArray) -> Adjacency:
        ret = self._c.times_farray(other._c if other is not None else None)
        return Adjacency._from_capi(ret)

    def equal(self, other: Adjacency) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: Adjacency) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def sum(self, ) -> None:
        ret = self._c.sum()
        return ret

    def where(self, value: Any) -> List:
        ret = self._c.where(value)
        if ret is None: return None
        return List(ret)

    def flip(self, axis: Any) -> Adjacency:
        ret = self._c.flip(axis)
        return Adjacency._from_capi(ret)

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __len__(self):
        return self.size()

    def __mul__(self, other):
        """Operator overload for *"""
        if hasattr(other, "_c") and type(other).__name__ == "EqualsFarray":
            return self.times_equals_farray(other)
        if hasattr(other, "_c") and type(other).__name__ in ["FArrayDouble", "FArrayInt", "FArray"]:
            return self.times_farray(other)
        return NotImplemented

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Adjacency):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Adjacency):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/src/physics/config/core/adjacency.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class type c_adjacency_t = object
  method raw : unit ptr
end
class c_adjacency (h : unit ptr) : c_adjacency_t = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.adjacency_destroy raw_val;
    Error_handling.raise_if_error ()
  ) self
end

module Adjacency = struct
  type t = c_adjacency

  let copy (handle : t) : t =
    Error_handling.read handle (fun () ->
      let ptr = Capi_bindings.adjacency_copy handle#raw in
      Error_handling.raise_if_error ();
      new c_adjacency ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.adjacency_from_json_string (Falcon_string.of_string json) in
    Error_handling.raise_if_error ();
    new c_adjacency ptr

  let make (data : int) (shape : int) (ndim : int) (indexes : Connections.Connections.t) : t =
    Error_handling.read indexes (fun () ->
      let ptr = Capi_bindings.adjacency_create data (Unsigned.Size_t.of_int shape) (Unsigned.Size_t.of_int ndim) indexes#raw in
      Error_handling.raise_if_error ();
      new c_adjacency ptr
    )

  let equal (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.adjacency_equal handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.adjacency_not_equal handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_to_json_string handle#raw in
      Error_handling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let indexes (handle : t) : Connections.Connections.t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_indexes handle#raw in
      Error_handling.raise_if_error ();
      new Connections.c_connections result
    )

  let getTruePairs (handle : t) : Listpairsizetsizet.ListPairSizeTSizeT.t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_get_true_pairs handle#raw in
      Error_handling.raise_if_error ();
      new Listpairsizetsizet.c_listpairsizetsizet result
    )

  let size (handle : t) : int =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_size handle#raw in
      Error_handling.raise_if_error ();
      Unsigned.Size_t.to_int result
    )

  let dimension (handle : t) : int =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_dimension handle#raw in
      Error_handling.raise_if_error ();
      Unsigned.Size_t.to_int result
    )

  let shape (handle : t) (out_buffer : int) (ndim : int) : int =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_shape handle#raw (Unsigned.Size_t.of_int out_buffer) (Unsigned.Size_t.of_int ndim) in
      Error_handling.raise_if_error ();
      Unsigned.Size_t.to_int result
    )

  let data (handle : t) (out_buffer : int) (numdata : int) : int =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_data handle#raw out_buffer (Unsigned.Size_t.of_int numdata) in
      Error_handling.raise_if_error ();
      Unsigned.Size_t.to_int result
    )

  let timesEqualsFarray (handle : t) (other : Farrayint.FArrayInt.t) : unit =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.adjacency_times_equals_farray handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let timesFarray (handle : t) (other : Farrayint.FArrayInt.t) : t =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.adjacency_times_farray handle#raw other#raw in
      Error_handling.raise_if_error ();
      new c_adjacency result
    )

  let sum (handle : t) : int =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_sum handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let where (handle : t) (value : int) : Listlistsizet.ListListSizeT.t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_where handle#raw value in
      Error_handling.raise_if_error ();
      new Listlistsizet.c_listlistsizet result
    )

  let flip (handle : t) (axis : int) : t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.adjacency_flip handle#raw (Unsigned.Size_t.of_int axis) in
      Error_handling.raise_if_error ();
      new c_adjacency result
    )

end